trunk/source/TraceAbstraction/src/de/uni_freiburg/informatik/ultimate/plugins/generator/traceabstraction/BasicCegarLoop.java

```diff
@@ -283,17 +283,21 @@ protected final void readInitialProof() throws AutomataLibraryException {
 		refineAbstraction();
 	}
 
+	protected abstract List<?> getControlLocationsFromCounterexample(final IRun<L, ?> run);
+
 	@Override
 	protected Pair<LBool, IProgramExecution<L, Term>> isCounterexampleFeasible()
 			throws AutomataOperationCanceledException {
 
 		IStatisticsDataProvider refinementEngineStats = null;
-		final ITARefinementStrategy<L> strategy = mStrategyFactory.constructStrategy(getServices(), mCounterexample,
-				mAbstraction, new SubtaskIterationIdentifier(mTaskIdentifier, getIteration()),
-				mPredicateFactoryInterpolantAutomata, getPreconditionProvider(), getPostconditionProvider());
+		final var locations = getControlLocationsFromCounterexample(mCounterexample);

```

Comment:
No. The states of the `IRun` should not be used for this (see second bullet point under "Motivation" above).

Reply:
Hm, I still dont understand it, in particular in combination with the implementation in NwaCegarLoop, which does exactly what you are saying it should not do.

Reply:
One should not use the states of the difference automaton _verbatim_, because they contain the interpolant automaton states as well as the program control flow.

The point of this method is to extract only the part of the run state sequence that reflects the control flow in the program, and throw away everything else.
- The implementation in `NwaCegarLoop` computes the set of all Icfg locations of all active threads at each point in the run.
  + Previously, the trace checks implemented something else: they took just one of the Icfg locations at each point, from one of the threads.
 - In `CegarLoopForPetriNet`, the computation is based on the markings of the Petri net, etc.

Reply:
:+1: